### ISSUE TITLE
loop counter false positive

### ISSUE THREAD

**Issue by frewsxcv** [Manishearth/rust-clippy#384]
```rust
#![feature(plugin)]

#![plugin(clippy)]


fn main(){
    let mut i = 0;

    for _ in 1..10 {
        i += 1;
    }

    println!("{}", i);
}
```

results in

```
src/lib.rs:9:5: 11:6 warning: the variable `i` is used as a loop counter. Consider using `for (i, item) in 1..10.enumerate()` or similar iterators, #[warn(explicit_counter_loop)] on by default
src/lib.rs: 9     for _ in 1..10 {
src/lib.rs:10         i += 1;
src/lib.rs:11     }
src/lib.rs:9:5: 11:6 note: in this expansion of for loop expansion
src/lib.rs:9:5: 11:6 help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#explicit_counter_loop
```

From what I understand, it is not possible to replicate the behavior of my program using `enumerate`, so it shouldn't be suggested if the loop counter value is used *outside* the loop body

**Comment by Manishearth** [comment]
cc @birkenfeld

**Comment by llogiq** [comment]
Piggybacking on this, I think I found another false positive when the indexing isn't directly done on the loop counter, e.g. (from rust-bio)

    let mut rank: Vec<usize> = iter::repeat(0).take(n).collect();
    for r in 0..n {
        rank[pos[r]] = r;
    }

**Comment by Manishearth** [comment]
Can't that be done with `for (p,r) in pos.iter().enumerate()`?

**Comment by llogiq** [comment]
You're right, the solution is: `for (p, r) in pos.iter().enumerate() { rank[p] = r }`. I was a bit addled by the warning message.

**Comment by Manishearth** [comment]
Doesn't the lint suggest that?

**Comment by llogiq** [comment]
Re-reading the message, it did. I must've misread it.
